### ISSUE TITLE
Self hosting on Unraid

### ISSUE THREAD

**Issue by SavageAUS** [Peergos/Peergos#915]
This project looks great and I can certainly find a use for it but due to the lack of an unraid template or this project being on dockerhub I can not use it unless I use a VM and that would use more resources. 
Can you please add this project to docker hub so I can create an unraid template for it.

**Comment by Olprog59** [comment]
Hello,
I discovered this GitHub yesterday. And I'm also interested in testing it and putting it on Unraid.
I'm in the process of mounting this on a Docker to see if it's feasible.

**Comment by ianopolous** [comment]
We'd happily accept a PR for a docker file, and add CI to build it.

**Comment by jonaswre** [comment]
I think i've asked that before, but you didn't wanted to host on docker hub.

**Comment by ianopolous** [comment]
I think if people are willing to trust docker hub and we can get it all building from the CI here then that's fine.

**Comment by jonaswre** [comment]
I agree and if you don't you could still build it your self. I think there needs to be some though put into what the docker image is able to do.

The image should let you run your instance standalone (available publicly/and private) and mirrored. 
Maybe provide the ability to use a S3 storage backend. 

It might be cool to have an option to fuse mount inside of the container just so you might mitigate the trouble I had with fuse mounting on windows. This could also let you integrate peergos in legacy processes like a scanner. Or any application that ingests something from a device. 

Docker Images should be available for arm64 and regular x64

Anything I'm missing? 

Any recommendations for a hardend base image?

**Comment by ianopolous** [comment]
My understanding of docker is very limited, but I think a minimal version needs to open the right ports for the api and ipfs (4001 and 8000) and mount a persistent volume which is used for the peergos dir (config, logs, sqlite and blockstore). 

Extensions could include S3 blockstore, postgres, fuse, nginx with TLS cert management.

I think an openjdk 18 base image should be fine. https://hub.docker.com/_/openjdk

**Comment by jonaswre** [comment]
What do you mean by extension? 
Other image that use peergos Base Image? 

The usual way, or at least the one I encounter the most, is to provide the configuration using Environment Variable so that when the container starts it's already configured. You don't want to encounter a situation where your container needs to be configured after first start. 

So you would have a Environment variable 

FUSE_MOUNT=true
FUSE_USERNAME=username
FUSE_PASSWORT=password

S3_HOST=
S3_BUCKET=
S3_SECRET=

POSTGRES_HOST=
POSTGRES_USERNAME=
POSTGRES_PASSWORD=
POSTGRES_DATABASE=

And I'm not sure if having nginx on board is best practice for a container. A reverse proxy is likely already present In larger deployments. And in Smaller ones we could provide a docker_compose that has one configured.

**Comment by ianopolous** [comment]
I guess in docker you can just change the run command if you want to use one image to do different things? Sure that sounds fine. 

We support environment vars, but they have to match the arg name at the moment. We may need to update that to support env vars with the dashes and dots replaced by _

**Comment by gsaslis** [comment]
@ianopolous assuming I would open a PR with a Dockerfile, what would be a good smoke test / health check to verify that the Peergos.jar starts correctly (so we can verify that the app has been packaged correctly in the container image) ? Is there, for example, some `/status` API endpoint that returns a `200 OK`, or something along those lines ?

**Comment by ianopolous** [comment]
That's awesome, @gsaslis ! Thank you. For status, we tend to use GET /public/peergos/releases which tests that it is up and running, and also able to reach the PKI server. It should return a HTTP redirect to a secret link to that (public) folder. For something that doesn't test external connectivity, you can just use GET /

**Comment by gsaslis** [comment]
PR opened here: https://github.com/Peergos/web-ui/pull/485

**Comment by gsaslis** [comment]
@ianopolous ack, yes, I'll make some time for that this week. 👍🏼

**Comment by gsaslis** [comment]
@ianopolous coming back to this now: 

I found https://github.com/Peergos/web-ui?tab=readme-ov-file#container-image-docker 

Were you referring to something else?

**Comment by gsaslis** [comment]
p.s. here is how I am deploying the webdav bridge, in my k3s homelab cluster, in case that also helps.

**Comment by ianopolous** [comment]
I think this is the first command: docker pull ghcr.io/peergos/web-ui:master

**Comment by ianopolous** [comment]
I've made a start with https://github.com/Peergos/Peergos/blob/master/README.md#usage---self-hosting-with-docker

Does that docker command work for a server with a public domain name?

**Comment by PhysicsIsAwesome** [comment]
> I've made a start with https://github.com/Peergos/Peergos/blob/master/README.md#usage---self-hosting-with-docker
> 
> Does that docker command work for a server with a public domain name?

I can't get it to work with

`docker run --volume $(PEERGOS_PATH):/opt/peergos/data peergos:dev daemon -domain 0.0.0.0 -public-domain localhost:8000 -log-to-console true`

Image `peergos:dev` is not found. Did you actually mean `ghcr.io/peergos/web-ui:master` like in the `docker pull` command? Is this the production release image or just a snapshot of the master branch?

But even `docker run --volume $(PEERGOS_PATH):/opt/peergos/data ghcr.io/peergos/web-ui:master daemon -domain 0.0.0.0 -public-domain peergos.mydomain.com -log-to-console true` does not seem to work.

Logs show that it does a long setup and after that shows that the server should now be reachable. But I can't access it. Neither with Caddy in front of it nor directly. It's unclear what the `-domain` flag does.

This is how my `compose.yaml` looks: 
```
services:
  peergos:
    image: ghcr.io/peergos/web-ui:master
    restart: always
    command: >
      daemon
      -domain 0.0.0.0
      -public-domain peergos.mydomain.com
      -log-to-console true
    expose:
      - "8000"
    volumes:
      - ./data:/opt/peergos/data
      - /etc/resolv.conf:/etc/resolv.conf:ro
    networks:
      rproxy-peergos:
        ipv4_address: "172.20.9.3"
    runtime: runsc  # gvisor systrap
    security_opt:
      - label=disable # disabling needed for runsc

networks:
  rproxy-peergos:
    external: true
```

Network "rproxy-peergos" is connected to a TLS-terminating Caddy reverse proxy container.

Caddyfile: 
```
peergos.mydomain.com, *.peergos.mydomain.com {
	reverse_proxy 172.20.9.3:8000
}
```

**Comment by ianopolous** [comment]
Excellent. "domain" arg was a terrible name. So we renamed it to "listen-host"

**Comment by ianopolous** [comment]
@PhysicsIsAwesome Did you change "-domain" to "-listen-host"?

**Comment by PhysicsIsAwesome** [comment]
> [@PhysicsIsAwesome](https://github.com/PhysicsIsAwesome) Did you change "-domain" to "-listen-host"?

Thx. Seems like changes to `daemon` within `command` don't go into effect once the server had been spin up one time.  I had to manually change it in the Peergos' `config` file. I can now access it 👍 

I tried to register an account on my instance, however on the step of "logging in" it showed "Error: Null" in the browser in a red notification, both for newly registered accounts on this instance and when entering my credentials of my peergos.net account. Container logs did not show any explicit errors.

In the nginx example config, you gave the `/.well-known` location special treatment and passed it to port 8888. Do you know if this is also needed for Caddy or is this something only needed with nginx?

**Comment by ianopolous** [comment]
> when entering my credentials of my peergos.net account.

The default is to not allow logins for users from other servers on public instances. Were there any console errors for the local login? What browser is this?

The .well-known pass through was for letsencrypt certbot. I don't know about caddy.

**Comment by PhysicsIsAwesome** [comment]
> What browser is this?

Brave [Version 1.76.82 Chromium: 134.0.6998.178 (Official Build) (64-Bit)](https://brave.com/latest/) . Logins on peergos.net with a peergos.net account work flawless with the same browser.

> Were there any console errors for the local login?

Yes:
```
GET       vendor.js:60 
https://peergos.mydomain.com/peergos/v0/mutable/getPointer?owner=onerandomstring&writer=anotherrandomstring 405 (Method Not Allowed)
```
occured multiple times. After that:
```
java.lang.JsException: Error: null
peergoslib.nocache.js:14268 	at Unknown.onload(https://peergos.mydomain.com/js/vendor.js)
```

> The .well-known pass through was for letsencrypt certbot.

Am I right to assume that certbot is not part of the Peergos docker container and thus there is no port 8888 available there?

**Comment by ianopolous** [comment]
the http 405 means it is blocking the request because it is not a http post. For a public server you can allow these with -public-server true

Yes, we don't include certbot.

**Comment by PhysicsIsAwesome** [comment]
> For a public server you can allow these with -public-server true

For now it is only available inside an internal network. Should I nevertheless set `-public-server true`?
When setting it, this occurs inside the container logs:
```
peergos-1  | Apr 07, 2025 4:05:41 PM org.peergos.util.HttpUtil replyError
peergos-1  | WARNING: Couldn't resolve SOMERANDOMSTRING
peergos-1  | java.lang.IllegalStateException: Couldn't resolve SOMERANDOMSTRING
peergos-1  |    at org.peergos.net.APIHandler.handleCallToAPI(APIHandler.java:296)
peergos-1  |    at org.peergos.net.Handler.handle(Handler.java:30)
peergos-1  |    at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Unknown Source)
peergos-1  |    at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(Unknown Source)
peergos-1  |    at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Unknown Source)
peergos-1  |    at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(Unknown Source)
peergos-1  |    at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Unknown Source)
peergos-1  |    at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
peergos-1  |    at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
```
Pressing sign in does nothing now in the browser and console.

Oh and there are some errors I might have missed, because they already occured when loading the site and I initially recorded console only after pressing login. These happen already when I load the website:
```
Error with Permissions-Policy header: Unrecognized feature: 'interest-cohort'.
```

```
vendor.js:1766 Uncaught (in promise) NotSupportedError: Failed to execute 'importKey' on 'SubtleCrypto': Algorithm: Unrecognized name
    at vendor.js:1766:22
```

**Comment by PhysicsIsAwesome** [comment]
> So it must be able to dial outside to the public internet.

I can resolve hosts and curl websites from inside the container without problems. Should I test something specific?

**Comment by PhysicsIsAwesome** [comment]
Is this what you mean? `INFO: ... finished updating pki mirror state.` `INFO: pki up-to-date`

**Comment by PhysicsIsAwesome** [comment]
I created a new account and with this account it works to login, I even shared some test data with an account on the peergos.net instance. Amazing :))

**Thank you so much for your fast help @ianopolous !**

I don't know why the other two accounts I created before making the changes you suggested did not work for logging in, despite them still being existent and explicitly mentioned in the container logs. Is there anyway I can delete them, so they don't block these usernames?

Would you recommend to switch to Postgres instead of Sqlite?

Reading through `peergos daemon -help` it seems like it is possible to mirror a user's data. Let's say I want to mirror one of my peergos.net's account data on my home server for faster access and backup, is that possible and how?
